File: 15-3sum/3sum.java
class Solution {
    
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        for(int i = 0;i < nums.length;i++)
        {
            if (i > 0 && nums[i] == nums[i - 1]) 
            {
                continue;
            }
            int f = i;
            int j = i+1;
            int k = nums.length-1;
            while(j < k)
            {
                List<Integer> l = new ArrayList<>();
                int sum = nums[f] + nums[j] + nums[k];
                if(sum == 0)
                {
                    l.add(nums[f]);
                    l.add(nums[j]);
                    l.add(nums[k]);
                    res.add(l);
                    j++;
                    k--;

                    while (j < k && nums[j] == nums[j - 1]) {
                        j++;
                    }
                }
                else if (sum > 0)
                {
                    k--;
                }
                else
                {
                    j++;
                }
            }
        }
        return res;
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        lis = []
        nums = sorted(nums)

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            f = i
            r = i+1
            l = len(nums)-1
            while(r < l):
                su = nums[f] + nums[r] + nums[l]
                lo = []
                if su == 0:
                    lo.append(nums[f])
                    lo.append(nums[r])
                    lo.append(nums[l])
                    lis.append(lo)
                    r += 1
                    l -= 1
                    
                    while(r < l and nums[r] == nums[r-1]):
                        r += 1

                    while(r < l and nums[l] == nums[l+1]):
                        l -= 1                    

                elif su > 0:
                    l -= 1

                else:
                    r += 1

                    

        return lis

